flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/ShuffleOperator.java

```diff
@@ -0,0 +1,132 @@
+import org.apache.flink.runtime.operators.coordination.OperatorEvent;
+import org.apache.flink.runtime.operators.coordination.OperatorEventGateway;
+import org.apache.flink.runtime.operators.coordination.OperatorEventHandler;
+import org.apache.flink.runtime.state.StateInitializationContext;
+import org.apache.flink.runtime.state.StateSnapshotContext;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperator;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+
+/**
+ * Shuffle operator can help to improve data clustering based on the key.
+ *
+ * <p>It collects the data statistics information, sends to coordinator and gets the global data
+ * distribution weight from coordinator. Then it will ingest the weight into data stream(wrap by a
+ * class{@link ShuffleRecordWrapper}) and send to partitioner.
+ */
+class ShuffleOperator<T, K> extends AbstractStreamOperator<ShuffleRecordWrapper<T, K>>
+    implements OneInputStreamOperator<T, ShuffleRecordWrapper<T, K>>, OperatorEventHandler {
+
+  private static final long serialVersionUID = 1L;
+
+  private final KeySelector<T, K> keySelector;
+  private final OperatorEventGateway operatorEventGateway;
+  // key is generated by applying KeySelector to record
+  // value is the times key occurs
+  // TODO: support to store statistics for high cardinality cases
+  private transient DataStatistics<K> localDataStatistics;
+  private transient DataStatistics<K> globalDataStatistics;
+  private transient ListState<DataStatistics<K>> globalDataStatisticsState;
+  private transient DataStatisticsFactory<K> dataStatisticsFactory;
+
+  public ShuffleOperator(
+      KeySelector<T, K> keySelector,
+      OperatorEventGateway operatorEventGateway,
+      DataStatisticsFactory<K> dataStatisticsFactory) {
+    this.keySelector = keySelector;
+    this.operatorEventGateway = operatorEventGateway;
+    this.dataStatisticsFactory = dataStatisticsFactory;
+  }
+
+  @VisibleForTesting
+  ListStateDescriptor<DataStatistics<K>> generateGlobalDataDistributionWeightDescriptor() {
+    return new ListStateDescriptor<>(
+        "globalDataStatisticsState", TypeInformation.of(new TypeHint<DataStatistics<K>>() {}));
+  }
+
+  @Override
+  public void initializeState(StateInitializationContext context) throws Exception {
+    localDataStatistics = dataStatisticsFactory.createDataStatistics();
+    globalDataStatisticsState =
+        context
+            .getOperatorStateStore()
+            .getListState(generateGlobalDataDistributionWeightDescriptor());
+
+    if (globalDataStatisticsState.get() != null
+        && globalDataStatisticsState.get().iterator().hasNext()) {
+      globalDataStatistics = globalDataStatisticsState.get().iterator().next();
+    } else {
+      globalDataStatistics = dataStatisticsFactory.createDataStatistics();
```

Comment:
this is a cold start problem (no state, rescale, etc.). does it make sense to create an empty Statistics object or we should keep it null to indicate cold start problem?